### DISCUSSION TITLE
Mastodon as OAuth2 Client (NOT the provider)

### DISCUSSION THREAD

**Discussion by AustinFoss** [mastodon/mastodon#19813] [General]
Hello! I've been looking over the documentation and searching for information regarding OAuth support.

What I found:
- https://docs.joinmastodon.org/spec/oauth/
- https://docs.joinmastodon.org/methods/apps/oauth/
- https://docs.joinmastodon.org/client/token/

From what I can see this is all to enable the Mastodon server as an 0Auth provider rather than as a client. If I wanted to sign into Mastodon with my self hosted Gitlab/Gitea for instance is this something that is, or will be, supported?  

I did find this shortly after posting: https://docs.joinmastodon.org/admin/config/#external-authentication

So it seems `OAUTH_REDIRECT_AT_SIGN_IN` is something that's built in, maybe, but just not yet documented?

**Comment by AustinFoss** [comment]
Hmmmm, well I'm not sure it is built in anymore after finding this in the change log file, seems it was removed in favor of `OMNIAUTH_ONLY`: https://github.com/mastodon/mastodon/blob/main/CHANGELOG.md#removed-3

Does this still make it possible for signing in with Gitlab or other providers?

**Comment by puyoxyz** [comment]
I was able to use OIDC, using config options I found in the comments of this: https://github.com/mastodon/mastodon/pull/16221

Seems like it will be documented here, but it's empty for now: https://docs.joinmastodon.org/admin/optional/sso/

Edit: Here's what mine looks like:

```
OIDC_ENABLED=true
OIDC_DISPLAY_NAME=dummy.cafe account
OIDC_ISSUER=https://dummy-cafe.eu.auth0.com/
OIDC_DISCOVERY=true
OIDC_SCOPE=openid,profile,email
OIDC_UID_FIELD=preferred_username
OIDC_CLIENT_ID=[redacted]
OIDC_CLIENT_SECRET=[redacted]
OIDC_REDIRECT_URI=https://toot.dummy.cafe/auth/auth/openid_connect/callback
OMNIAUTH_ONLY=true
```

**Reply by AustinFoss** [reply]
Oh awesome! thanks for the response, will give this spin sometime this week.

**Reply by alansill** [reply]
I'm getting set up to test this with an external UAuth2 (OIDC) provider. How should the OIDC_CLIENT_ID and ...SECRET be generated? Is it just like generation of secrets and/or the VAPID_... key pairs? 

In light of [this commit for remove support for OAUTH_REDIRECT_AT_SIGN_IN](https://github.com/mastodon/mastodon/pull/17287/commits/4030edde203290464e17bdbfba973a373be9e7d3), is it possible to enable the user to be able to select either to log in by external sign-on integration or by a local account?

**Reply by puyoxyz** [reply]
> I'm getting set up to test this with an external UAuth2 (OIDC) provider. How should the OIDC_CLIENT_ID and ...SECRET be generated? Is it just like generation of secrets and/or the VAPID_... key pairs?

Usually whatever OIDC provider you're using should let you make "Applications", which then have a Client ID and Secret

**Reply by puyoxyz** [reply]
> In light of [this commit for temove support for OAUTH_REDIRECT_AT_SIGN_IN](https://github.com/mastodon/mastodon/pull/17287/commits/4030edde203290464e17bdbfba973a373be9e7d3), is it possible to enable the user to be able to select either to log in by external sign-on integration or by a local account?

Yes, that's the default behavior

OMNIAUTH_ONLY=false to have both local accounts and SSO, OMNIAUTH_ONLY=true to have only SSO and no local accounts

**Reply by alansill** [reply]
Thanks! Others have pointed out the [Google OAUTH2 playground](https://developers.google.com/oauthplayground/) and I have found [this blog article from Okta useful](https://developer.okta.com/blog/2022/07/14/add-auth-to-any-app-with-oauth2-proxy).

**Comment by alansill** [comment]
Okay, I can make the OIDC configuration work and can create an account or sign in to the one created by OIDC, but it comes out with a random identifier string rather than mapping to an existing account. how can I control the account ID associated with a given OIDC sign-in? Is that something configurable about the Open ID exchange (i.e., should I configure the OIDC server to return a field like an email address) or is that something that needs to be configured locally on my Mastodon server?

**Reply by ineffyble** [reply]
If the account ID is present in the attributes being passed by the exchange, you need to set `OIDC_UID_FIELD` to the right field name. If it's not, then you'll need to make changes at the exchange first.

**Reply by alansill** [reply]
Seems like Mastodon with OIDC will need the ability to allow users to set a userID string to be associated with the account upon sign-in, rather than use the internal identifier provided by the OIDC during the exchange. Many times these are not designed to be seen in public and in at least one case I am testing, the subject returned as the unique identifier is longer than the field for the Mastodon userID and so can't be used directly.

**Reply by puyoxyz** [reply]
> Seems like Mastodon with OIDC will need the ability to allow users to set a userID string to be associated with the account upon sign-in, rather than use the internal identifier provided by the OIDC during the exchange. Many times these are not designed to be seen in public and in at least one case I am testing, the subject returned as the unique identifier is longer than the field for the Mastodon userID and so can't be used directly.

You can just set OIDC_UID_FIELD... if you don't have a username field already, that's a problem with the OIDC provider not with Mastodon

**Reply by jdkruzr** [reply]
heads up: I ran into this recently and had success setting OIDC_UID_FIELD to "preferred_username." my instance auths against Keycloak which in turn auths against Active Directory (really Zentyal Linux, but effectively the same).
